### PR TITLE
Feature/ruanpc/main prover

### PR DESCRIPTION
**Pull request by RUAN0007**

### CODE REVIEW

File: config/default.toml [hunk 1]
```diff
@@ -1,14 +1,14 @@
 [prover]
 round_no = 0
-batch_size = 8
-hyper_tree_size = 4
+batch_size = 1024
+hyper_tree_size = 64
 user_data_path = "/opt/data/zkpor/users/"
-num_of_tokens = 22
+num_of_tokens = 220
```
Comment (cliff0412) [inline config/default.toml:6]: we can put to prod.toml

File: crates/zk-por-core/src/bin/prover.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+struct Proof {
+    round_num: usize,
+    root_vd_digest: HashOut<F>,
+    proof: ProofWithPublicInputs<F, C, D>,
+}
+
+fn main() {
+    let cfg = ProverConfig::try_new().unwrap();
+    let trace_cfg: TraceConfig = cfg.log.into();
+    init_tracing(trace_cfg);
```
Comment (cliff0412) [inline crates/zk-por-core/src/bin/prover.rs:27]: let _g =

File: crates/zk-por-core/src/bin/prover.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+    const RECURSION_FACTOR: usize = 64;
```
Comment (cliff0412) [inline crates/zk-por-core/src/bin/prover.rs:29]: we may use HYPER_TREE_SIZE for consistency understanding

File: crates/zk-por-core/src/bin/prover.rs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    if cfg.prover.hyper_tree_size as usize != RECURSION_FACTOR {
+        panic!("The hyper_tree_size is not configured to be equal to 64 (Recursion_Factor)");
+    }
+    let batch_size = cfg.prover.batch_size as usize;
+    let asset_num = cfg.prover.num_of_tokens as usize;
+
+    // the path to dump the final generated proof
+    let mut bench_mode = true;
+    let args: Vec<String> = env::args().collect();
+    let arg1 = args.get(1).expect(
```
Comment (cliff0412) [inline crates/zk-por-core/src/bin/prover.rs:42]: as in pay 2 commitment. we use cli with --help. just for better user experience; we can change later

File: crates/zk-por-core/src/bin/prover.rs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    let cfg = ProverConfig::try_new().unwrap();
+    let trace_cfg: TraceConfig = cfg.log.into();
+    init_tracing(trace_cfg);
+
+    const RECURSION_FACTOR: usize = 64;
+    const BATCH_PROVING_THREADS_NUM: usize = 4;
+    const RECURSIVE_PROVING_THREADS_NUM: usize = 2;
+
+    if cfg.prover.hyper_tree_size as usize != RECURSION_FACTOR {
+        panic!("The hyper_tree_size is not configured to be equal to 64 (Recursion_Factor)");
```
Comment (cliff0412) [inline crates/zk-por-core/src/bin/prover.rs:34]: 64 can be changed to RECURSION_FACTOR

File: crates/zk-por-core/src/bin/prover.rs [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        let parser = FilesParser::new(FilesCfg {
+            dir: std::path::PathBuf::from_str(&cfg.prover.user_data_path).unwrap(),
+            batch_size: cfg.prover.batch_size,
+            num_of_tokens: cfg.prover.num_of_tokens,
+        });
+        parser.log_state();
+        account_reader = Box::new(parser);
+    }
+
+    // Hardcode three levels of recursive circuits, each branching out 64 children, with the last level with zk enabled.
```
Comment (cliff0412) [inline crates/zk-por-core/src/bin/prover.rs:68]: i will suggest add a TODO marker; we'd better not hard code

File: crates/zk-por-core/src/bin/prover.rs [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    tracing::info!(
+        "start to prove {} accounts with {} tokens, {} batch size",
+        account_reader.total_num_of_users(),
+        asset_num,
+        batch_size
+    );
+
+    let start = std::time::Instant::now();
+    let mut offset = 0;
+    let per_parse_account_num = 10 * batch_size;
```
Comment (cliff0412) [inline crates/zk-por-core/src/bin/prover.rs:88]: can use cpu_get

File: crates/zk-por-core/src/bin/prover.rs [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    let mut parse_num = 0;
+    let mut batch_proofs = vec![];
+    while offset < account_reader.total_num_of_users() {
+        parse_num += 1;
+        let accounts = account_reader.read_n_accounts(offset, per_parse_account_num);
+        let account_num = accounts.len();
+        assert_eq!(account_num % batch_size, 0);
```
Comment (cliff0412) [inline crates/zk-por-core/src/bin/prover.rs:95]: it is possible the last batch might not be a multiple of batch_size in actual case

File: crates/zk-por-core/src/circuit_registry/registry.rs [hunk 9]
```diff
@@ -17,14 +20,17 @@ use crate::{
...
 pub struct CircuitRegistry<const RECURSIVE_FACTOR: usize> {
     batch_circuit: (CircuitData<F, C, D>, Vec<AccountTargets>),
-    empty_batch_proof: ProofWithPublicInputs<F, C, D>,
-    recursive_circuits_and_empty_proofs: Vec<(
-        (CircuitData<F, C, D>, RecursiveTargets<RECURSIVE_FACTOR>),
-        ProofWithPublicInputs<F, C, D>,
-    )>,
+    // inner_vd => the verification circuit that verify the inner circuit
+    recurisve_circuits:
+        HashMap<HashOut<F>, (CircuitData<F, C, D>, RecursiveTargets<RECURSIVE_FACTOR>)>,
```
Comment (EkamSinghPandher) [inline crates/zk-por-core/src/circuit_registry/registry.rs:30]: typo

File: crates/zk-por-core/src/e2e.rs [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+                )
+                .as_str(),
+            );
+
+        let (recursive_circuit, recursive_targets) = circuit_registry
+            .get_recursive_circuit(&last_level_vd_digest)
+            .expect(format!("No recursive circuit found for level {}", level).as_str());
+        let subproof_len = last_level_proofs.len();
+
+        let mut subproof_batches: Vec<Vec<ProofWithPublicInputs<F, C, D>>> = last_level_proofs
```
Comment (cliff0412) [inline crates/zk-por-core/src/e2e.rs:68]: do not have to rearrange the proofs data structure to a vec of vec

**Review by cliff0412 — APPROVED** [review]

**Pull request merged**